components/script/dom/writablestreamdefaultwriter.rs

```diff
@@ -0,0 +1,490 @@
+        // Note: new promise created in `new_inherited`.
+        let ready_promise = self.ready_promise.borrow();
+        ready_promise.reject_native(&error.handle());
+        ready_promise.set_promise_is_handled();
+
+        // Set writer.[[closedPromise]] to a promise rejected with storedError.
+        // Set writer.[[closedPromise]].[[PromiseIsHandled]] to true.
+        // Note: new promise created in `new_inherited`.
+        let ready_promise = self.closed_promise.borrow();
+        ready_promise.reject_native(&error.handle());
+        ready_promise.set_promise_is_handled();
+
+        Ok(())
+    }
+
+    pub(crate) fn reject_closed_promise_with_stored_error(&self, error: &SafeHandleValue) {
+        self.closed_promise.borrow().reject_native(error);
+    }
+
+    pub(crate) fn set_close_promise_is_handled(&self) {
+        self.closed_promise.borrow().set_promise_is_handled();
+    }
+
+    pub(crate) fn set_ready_promise(&self, promise: Rc<Promise>) {
+        *self.ready_promise.borrow_mut() = promise;
+    }
+
+    pub(crate) fn resolve_ready_promise_with_undefined(&self) {
+        self.ready_promise.borrow().resolve_native(&());
+    }
+
+    pub(crate) fn resolve_closed_promise_with_undefined(&self) {
+        self.closed_promise.borrow().resolve_native(&());
+    }
+
+    /// <https://streams.spec.whatwg.org/#writable-stream-default-writer-ensure-ready-promise-rejected>
+    pub(crate) fn ensure_ready_promise_rejected(
+        &self,
+        global: &GlobalScope,
+        error: SafeHandleValue,
+        can_gc: CanGc,
+    ) {
+        let mut ready_promise = self.ready_promise.borrow_mut();
+        // If writer.[[readyPromise]].[[PromiseState]] is "pending", reject writer.[[readyPromise]] with error.
+        if ready_promise.is_pending() {
+            ready_promise.reject_native(&error);
+        } else {
+            // Otherwise, set writer.[[readyPromise]] to a promise rejected with error.
+            let promise = Promise::new(global, can_gc);
+            promise.reject_native(&error);
+            *ready_promise = promise;
+        }
+        // Set writer.[[readyPromise]].[[PromiseIsHandled]] to true.
+        ready_promise.set_promise_is_handled();
+    }
+
+    /// <https://streams.spec.whatwg.org/#writable-stream-default-writer-ensure-closed-promise-rejected>
+    pub(crate) fn ensure_closed_promise_rejected(
+        &self,
+        global: &GlobalScope,
+        error: SafeHandleValue,
+        can_gc: CanGc,
+    ) {
+        let mut closed_promise = self.closed_promise.borrow_mut();
+        // If writer.[[closedPromise]].[[PromiseState]] is "pending", reject writer.[[closedPromise]] with error.
+        if closed_promise.is_pending() {
+            closed_promise.reject_native(&error);
+        } else {
+            // Otherwise, set writer.[[closedPromise]] to a promise rejected with error.
+            let promise = Promise::new(global, can_gc);
+            promise.reject_native(&error);
```

Comment:
GC risk with active mutable borrow.

Reply:
Done, also in a few other place in the file.